### ISSUE TITLE
CholLazyTensor uses `inv_matmul` for `inv_quad` but it should use `_cholesky_solve`

### ISSUE THREAD

**Issue by rhaps0dy** [cornellius-gp/gpytorch#1967]
It's not taking advantage of its `self.root` tensor being lower triangular. Is `CholLazyTensor(lower_tril_mat)` now intended to be called as `CholLazyTensor(LowerTriangularLazyTensor(lower_tril_mat))` now? 

The `inv_matmul` used here:

https://github.com/cornellius-gp/gpytorch/blob/aa0ff73cbd065a3a29b893c89bdab5497da63298/gpytorch/lazy/chol_lazy_tensor.py#L61-L69

contrasts with the `_cholesky_solve` used here:

https://github.com/cornellius-gp/gpytorch/blob/aa0ff73cbd065a3a29b893c89bdab5497da63298/gpytorch/lazy/chol_lazy_tensor.py#L50-L59

This should not be a bug most of the time, but it gives me numerical errors.

**Comment by rhaps0dy** [comment]
How to reproduce this bug:
tensor([[ 0.7139],
        [-0.1197],
        [-2.1677],
        [-2.4531],
        [-2.4605]], grad_fn=<UnsqueezeBackward0>)

```
import torch
from gpytorch.lazy import CholLazyTensor

b = torch.tensor([[ 0.7139],
        [-0.1197],
        [-2.1677],
        [-2.4531],
        [-2.4605]])


chol = torch.tensor([[ 0.7964,  0.0000,  0.0000,  0.0000,  0.0000],
        [-0.2821,  0.0857,  0.0000,  0.0000,  0.0000],
        [-0.8392, -0.9878,  0.0926,  0.0000,  0.0000],
        [-1.0211,  0.5174, -1.3137,  0.7495,  0.0000],
        [ 0.5249, -1.2347,  0.4207,  0.3914,  1.0156]])


# Works well
CholLazyTensor(chol).inv_matmul(b)

# Should not break but it does
CholLazyTensor(chol).inv_quad(b)
```

This worked fine in Gpytorch 1.1.1.

**Comment by Balandat** [comment]
Hmm yeah looks like the assumption in the constructor of `CholLazyTensor` is that the arg is a `TriangularLazyTensor`. I think what we can do is check for whether that's the case, raise a `DeprecationWarning` if not and automatically convert the input. We should also be able to get rid of the `upper` kwarg in the constructor as we can read this out from the input (or infer it from the dense tensor while we are still in the deprecation stage. I'll try to put up a PR for this soon.

**Comment by rhaps0dy** [comment]
Thank you for merging a fix! I don't think it was slow :) . I was able to change my code to follow the new convention anyways.

**Comment by Lezcano** [comment]
I just randomly bumped into this issue. I think, if I've understood things correctly, that what you should be using here is `linalg.solve_triangular` (or the GPyTorch equivalent) if you want to take advantage of the triangular structure of the matrix. That should give you roughly a x3 speed-up over using a regular solve.

**Comment by rhaps0dy** [comment]
Hi Mario! If called on `LowerTriangularLazyTensor`, the function `inv_quad` already uses `solve_triangular`.

[Balandat's added PR](https://github.com/cornellius-gp/gpytorch/pull/1969) warns you (or errors? Can't remember) if the thing inside the cholesky lazy tensor is *not* typed as triangular, so now `solve_triangular` is always used when needed.

**Comment by Lezcano** [comment]
Oh, that's great to hear! Sorry for the noise!
